pkg/dependencies/dependents.go

```diff
@@ -0,0 +1,237 @@
+
+	if err != nil {
+		return nil, fmt.Errorf("error getting dependencies: %v", err)
+	}
+
+	for _, resp := range sboms.HasSBOM {
+		// Skip entries from "deps.dev" because they are inconsistent.
+		if resp.Origin == deps_dev.DepsCollector {
+			continue
+		}
+		// Iterate through the included dependencies of each SBOM.
+		for _, dependency := range resp.IncludedDependencies {
+			// Construct unique names for the dependency package and the package itself.
+			// TODO: Make the names actually unique, not just add "_".
+			depPkgName := dependency.DependencyPackage.Type + "_" + dependency.DependencyPackage.Namespaces[0].Namespace + "_" + dependency.DependencyPackage.Namespaces[0].Names[0].Name
+			pkgName := dependency.Package.Type + "_" + dependency.Package.Namespaces[0].Namespace + "_" + dependency.Package.Namespaces[0].Names[0].Name
+
+			var depPkgIds []string
+			pkgId := dependency.Package.Namespaces[0].Names[0].Versions[0].Id
+
+			if len(dependency.DependencyPackage.Namespaces[0].Names[0].Versions) == 0 {
+				findMatchingDepPkgVersionIDs, err := FindDepPkgVersionIDs(ctx, gqlClient, dependency.DependencyPackage.Type,
+					dependency.DependencyPackage.Namespaces[0].Namespace,
+					dependency.DependencyPackage.Namespaces[0].Names[0].Name, dependency.VersionRange)
+				if err != nil {
+					return nil, fmt.Errorf("error from FindMatchingDepPkgVersionIDs:%w", err)
+				}
+				depPkgIds = append(depPkgIds, findMatchingDepPkgVersionIDs...)
+			} else {
+				depPkgIds = append(depPkgIds, dependency.DependencyPackage.Namespaces[0].Names[0].Versions[0].Id)
+			}
+
+			for _, depPkgId := range depPkgIds {
+				// Map the IDs to their names.
+				idToName[depPkgId] = depPkgName
+				idToName[pkgId] = pkgName
+
+				// Skip "guac" files.
+				if dependency.DependencyPackage.Type == "guac" && dependency.DependencyPackage.Namespaces[0].Namespace == "files" {
+					continue
+				}
+
+				// First we need to find all the packages that have pkgName as a dependent
+
+				// Initialize a visited map and a queue for BFS to find all packages that have pkgName as a dependent.
+				visited := make(map[string]bool)
+				queue := []string{depPkgId}
+
+				// Perform BFS to mark visited nodes.
+				for len(queue) > 0 {
+					n := len(queue)
+
+					for i := 0; i < n; i++ {
+						node := queue[0]
+						queue = queue[1:]
+
+						if _, ok := visited[node]; ok {
+							continue
+						}
+						visited[node] = true
+
+						// Add dependency nodes of node to the queue.
+						queue = append(queue, dependencyEdges[node]...)
+					}
+				}
+
+				// Next we want to find all the packages that are dependencies of pkgName.
+				// We need to add them all to the dependencies of all nodes that have pkgName as a dependency.
+
+				// Reset the queue to find all packages that are dependencies of pkgName.
+				queue = []string{pkgId}
+
+				// Perform BFS to add pkgName and dependencies of it to all nodes that have pkgName as a dependency.
+				for len(queue) > 0 {
```

Comment:
Is it possible to remove one of these loops, in order to reduce the nesting depth of this BFS?

Reply:
We can’t remove these loops because we have to first get all the dependent packages, and then assign all the newly acquired dependency packages as the dependencies of the dependent packages.